### PR TITLE
Jetpack Agency Dashboard: implement site favorites and grouping for the agency dashboard

### PR DESCRIPTION
**Pull request by yashwin**
#### Proposed Changes

This PR adds favorites and grouping features to the agency dashboard

#### Testing Instructions

**Prerequisites**

1. Since this change is made specifically for agencies, you will have to set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/implement-site-favourite-and-grouping` and `yarn start-jetpack-cloud`.
2. Visit http://jetpack.cloud.localhost:3000/, and you'll be redirected to the `/dashboard`.
4. Verify that you are able to see 2 tabs - **All** & **Favorites** with the count.
5. Hover over any site and verify that you are able to see a `star` icon only when hovered.
6. Click on the `star` icon and verify that the site is set as favorite, you should see a notification for this.
7. Click on **View Favorites** in the success notification which is shown when the site is set as a favorite. 
8. Remove any site as a favorite and verify that the site is being removed from the favorite section, you should also see a notification.
9. Play around with the feature and verify everything works as expected. 
10. Switch to mobile view using the dev tool and verify the above steps work as expected.
11. On mobile view, the star icon when not set as a favorite should always show up.
12. From the `All Sites` tab, set any site as a favorite, click on `View Favorites` on the notification, and you should see that the `Favorites` tab should show an animation to highlight it.

**Screenshots**

**All** tab - Large Screen

Screenshot 2022-07-01 at 3 42 25 PM

**Favorites** tab - Large Screen

Screenshot 2022-07-01 at 3 42 32 PM

**All Sites** tab - Small Screen

Screenshot 2022-07-01 at 3 41 19 PM

Tabs dropdown - Small Screen

Screenshot 2022-07-01 at 3 41 50 PM

**Favorites** tab - Small Screen

Screenshot 2022-07-01 at 3 42 11 PM

**Screen Recordings**

Large Screen

https://user-images.githubusercontent.com/10586875/176882549-fe4797c0-a911-47f5-b2fc-90320e1fd2a1.mov

Small Screen

https://user-images.githubusercontent.com/10586875/176883189-4f0f7b66-f8ab-423f-899b-816ced914904.mov

Related to 1202076982646589-as-1202479055261985 & 1202076982646589-as-1202439683846931

### CODE REVIEW

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx
Comment (atanas-dev) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx]: I've started noticing this pattern of using separate `isSubmitting` state for requests when using react-query.
React Query already provides this for us in the form of `.isLoading`:
```
const { isLoading } = useMutation();
const { isLoading, isFetching } = useQuery();
```

Since in this case, we have 2 mutations, either of which could be loading then we can use something like this:
```
const setFavorite = useSetFavoriteSiteMutation( handleMutation() );
const removeFavorite = useRemoveFavoriteSiteMutation( handleMutation() );
const isSubmitting = setFavorite.isLoading || removeFavorite.isLoading;
```

This way we avoid having extra state that:
- duplicates already existing state
- forces us to manually keep it in sync inside success and error handlers via setIsSubmitting

Site-note:
We can use a single mutation (e.g. `useToggleFavoriteSiteMutation`) that accepts an argument whether the favorite status should be set or removed which can simplify things.

Comment (yashwin) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx]: @atanas-dev - Thanks for your suggestion! I have updated it accordingly and used a single mutation to toggle favorites. 🙂 👍

Comment (atanas-dev) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx]: Happy to help :)

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/style.scss
Comment (rcanepa) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/style.scss]: We need to fix the indentation of these two lines.

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/test/site-content.tsx
Comment (rcanepa) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/test/site-content.tsx:6]: Couldn't we wrap a component higher in the component chain to avoid wrapping many components?

Comment (yashwin) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/test/site-content.tsx:6]: Thanks for the review, I have addressed your comments.

**Comment by jeffgolenski** [comment]
Thanks @yashwin - This is awesome to see. Other than the loading delays which I think you all are addressing, I think we're almost there. Just a couple items of feedback:

* I made a mistake in the dev notes with the background colors (sorry!). The bottom part of the back should have a background set to white at 50% opacity (or #FAFBFB) - whichever method you prefer. Attaching the image:

Dashboard favorites - page attributes

* The star icon appears to be different than the one we should be using from the [WordPress design library](5InhO7vz5wViXzZP9afcIe-fi-5415%3A21381), which has much more subtle rounded points. 

Screen Shot 2022-07-05 at 2 27 17 PM

We should be pulling from that icon / component library, if we aren't already.

**Comment by vitozev** [comment]
> [...] The bottom part of the back should have a background set to white at 50% opacity (or #FAFBFB) - whichever method you prefer. 

This issue is now resolved, here's a quick screencast:

https://user-images.githubusercontent.com/1749918/177552820-b4e91cd8-2f43-493a-b679-df8972751381.mp4

Does it look alright, @jeffgolenski?

> The star icon appears to be different than the one we should be using from the [WordPress design library](https://github.com/Automattic/wp-calypso/pull/5InhO7vz5wViXzZP9afcIe-fi-5415%3A21381), which has much more subtle rounded points.

It looks like we are using the right two icons:

```
<Gridicon size={ 18 } icon={ isFavorite ? 'star' : 'star-outline' } />
```

Could you please show the difference between the design library and the dashboard?

**Comment by jeffgolenski** [comment]
The background change looks ace, @vitozev, thank you for the that.

RE the icons.

Traditionally, gridicons have been loaded into calypso. But we're implementing the icons from the WordPress Design Library (gutenberg icons) for all Jetpack products moving forward. You can see the star icons in this repo below

https://automattic.github.io/jetpack-storybook/?path=/story/gutenberg_icons-icon--library

If there aren't already loaded in, it's not a big deal. but I was under the impression we started using those a long while ago. Thanks my friend!

**Comment by vitozev** [comment]
Thanks for the explanation, @jeffgolenski!

I've updated the icons:

https://user-images.githubusercontent.com/1749918/177614962-fbb20f0d-0a54-4f15-be27-00ab7dc58ff5.mp4

Let me know what you think.

**Comment by jeffgolenski** [comment]
@vitozev it's beautiful! thanks for the quick turn around. is that gutenberg icon set already loaded into calypso or did you have to do some tinkering?

**Review by rcanepa — APPROVED** [review]
Amazing implementation, @yashwin!

After manually testing these changes for a few days, I'm confident they are ready to be shipped or close to it.

From a code perspective, I didn't find anything concerning/in need of a change, so I'm approving the PR. However, it would be good to have another pair of eyes before we merge it.

**Comment by vitozev** [comment]
> @vitozev it's beautiful! thanks for the quick turn around. is that gutenberg icon set already loaded into calypso or did you have to do some tinkering?

@jeffgolenski, they were already loaded in Calypso, and I just used the ones we need here.

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts
Comment (manzoorwanijk) [inline client/data/agency-dashboard/use-fetch-dashboard-sites.ts:17]: Why not simply this?

```suggestion
		show_only_favorites: Boolean( filter.showOnlyFavorites ),
```

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx
Comment (manzoorwanijk) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx]: Is it a typo here?

Should it be `highlightTab` and `setHighlightTab`?

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-card/index.tsx
Comment (manzoorwanijk) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-card/index.tsx]: Can we wrap this in `useCallback`?

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx
Comment (manzoorwanijk) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx]: This to me seems like an expensive computation and the result is a non-static value on each render which will result in further un-necessary re-renders. So, it's better to wrap it in `useMemo`.

<details>
<summary>Here is the suggestion</summary>

```suggestion
	const navItems = useMemo(
		() =>
			[
				{
					key: 'all',
					label: isMobile ? translate( 'All Sites' ) : translate( 'All' ),
				},
				{
					key: 'favorites',
					label: translate( 'Favorites' ),
				},
			].map( ( navItem ) => {
				const isFavorite = navItem.key === 'favorites';
				return {
					...navItem,
					count: ( isFavorite ? data?.totalFavorites : data?.total ) || 0,
					selected: isFavorite ? filter.showOnlyFavorites : ! filter.showOnlyFavorites,
					path: `${ basePath }${ isFavorite ? '/favorites' : '' }${ search ? '?s=' + search : '' }`,
					onClick: () => {
						setHightLightTab( false );
						dispatch(
							recordTracksEvent( 'calypso_jetpack_agency_dashboard_tab_click', {
								nav_item: navItem.key,
							} )
						);
					},
					children: navItem.label,
				};
			} ),
		[
			data?.total,
			data?.totalFavorites,
			dispatch,
			filter.showOnlyFavorites,
			isMobile,
			search,
			translate,
		]
	);
```

You of course need to import `useMemo`.
</details>

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx
Comment (manzoorwanijk) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx]: This fragment is unnecessary.

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx
Comment (manzoorwanijk) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx]: Can we have an `aria-label` for this button for what it does?

**Review by manzoorwanijk — COMMENTED** [review]
Looks good. Just have a few suggestions related to code improvements.

**Review by manzoorwanijk — COMMENTED** [review]
Also, with regards to using `left` and `right` in CSS rules, it's better to use [RTL aware](https://hacks.mozilla.org/2015/10/building-rtl-aware-web-apps-websites-part-2/) alternatives:
`margin-left` -> `margin-inline-start`
`margin-right` -> `margin-inline-end`

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts
Comment (vitozev) [reply client/data/agency-dashboard/use-fetch-dashboard-sites.ts:17]: Thanks! Fixed in 09ad234a5ec3a6048bbe2b7471b077ea831c70ac

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx
Comment (vitozev) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx]: Were you able to find where this logic is being used? I removed it and didn't find any regression (both desktop and mobile).

This code was removed in: 09ad234a5ec3a6048bbe2b7471b077ea831c70ac

I'll tag @yashwin to take a look when he's back, but I think it might be some WIP work.

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx
Comment (vitozev) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx]: Thanks! Fixed in 914341990cbb2cb306bbe6ba6d10040b0b954483

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx
Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx]: Yes, <s>I didn't find any regression either. Looks good now.</s> Thanks

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx]: Sorry, just noticed a regression after you removed that logic. 

It no longer shows all the sites, it only shows Favorite sites on mobile view.

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-card/index.tsx
Comment (vitozev) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-card/index.tsx]: Fixed in: 44f4bc6fad8841a6816d5b7c41c06a8a85a429ea

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx
Comment (vitozev) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx]: Fixed in 44f4bc6fad8841a6816d5b7c41c06a8a85a429ea

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx
Comment (vitozev) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx]: Thanks! The commit was reverted and the typo was fixed in a8d2dbe4dd7feacdd78e61d61878f03f45e9d9e1

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts
Comment (vitozev) [reply client/data/agency-dashboard/use-fetch-dashboard-sites.ts:17]: I ended up reverting this one -- it looks like we've done it like this because of the way PHP check for that query parameter works.

**Comment by vitozev** [comment]
@manzoorwanijk, thanks for all your great feedback here! I have addressed all of it.

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts
Comment (manzoorwanijk) [reply client/data/agency-dashboard/use-fetch-dashboard-sites.ts:17]: Aha! got it. Thanks

**Review by manzoorwanijk — APPROVED** [review]
Looks great now. Thank you for making the suggested changes.

**Comment by rcanepa** [comment]
> Also, with regards to using `left` and `right` in CSS rules, it's better to use [RTL aware](https://hacks.mozilla.org/2015/10/building-rtl-aware-web-apps-websites-part-2/) alternatives: `margin-left` -> `margin-inline-start` `margin-right` -> `margin-inline-end`

I wasn't aware of these rules. Thanks for sharing them!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7426183

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

**Comment by keoshi** [comment]
Hey, all! I'm getting an error when trying to favorite a site for the first time:

> Cannot read properties of undefined (reading 'total_favorites')

image

It seems to stem from this line:

https://github.com/Automattic/wp-calypso/blob/f6515d0f646b1d6718fe315f0247e0280cb37be9/client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-set-favorite/index.tsx#L106

**Comment by jeffgolenski** [comment]
Great work y'all super excited to see this. 

I was wondering if we could make a few enhancements to the empty state for the favorites section:

Currently exists:

Screen Shot 2022-07-11 at 1 56 19 PM



Proposed:

Screen Shot 2022-07-11 at 1 57 31 PM

* Change "No active sites" to "You have no favorite sites yet"  (they do have active sites, just no favorites, we should be clear here)
* Remove search and filter component. In the empty state there's nothing to search or filter by
* May be unrelated, but there's also a scroll bar that exists on both the all sites and favorite views, even though the content doesn't need it. 

https://user-images.githubusercontent.com/214813/178328628-c4fa3113-1c11-4272-9333-ddf7892eff03.mov



Thanks y'all!

**Comment by jeffgolenski** [comment]
Also, with the addition of favorites, I'm noticing that the site domain no longer aligns with the "site" in the header column
Screen Shot 2022-07-11 at 2 08 09 PM

**Comment by yashwin** [comment]
> Hey, all! I'm getting an error when trying to favorite a site for the first time:
> 
> > Cannot read properties of undefined (reading 'total_favorites')

@keoshi - Thanks for reporting the bug 

PR is up for this: https://github.com/Automattic/wp-calypso/pull/65483

**Comment by yashwin** [comment]
> * Change "No active sites" to "You have no favorite sites yet"  (they do have active sites, just no favorites, we should be clear here)
> * Remove search and filter component. In the empty state there's nothing to search or filter by
> * May be unrelated, but there's also a scroll bar that exists on both the all sites and favorite views, even though the content doesn't need it.

Thanks for reporting these issues, @jeffgolenski 

We have created the tickets for these.

Regarding - `Remove search and filter component. In the empty state, there's nothing to search or filter by`, I think we can keep it since the empty state can appear when a filter is applied or when a user searches for a site. Please let me know your thoughts.

**Comment by jeffgolenski** [comment]
> Regarding - Remove search and filter component. In the empty state, there's nothing to search or filter by, I think we can keep it since the empty state can appear when a filter is applied or when a user searches for a site. Please let me know your thoughts.

@yashwin  I think those are 2 separate states to consider, but if it's an issue I suppose we could treat them similar. 

Ideally, we should have
1) If no favorite sites exist, hide search and filtering. Display "You don't have any favorites yet."
2) If favorite sites exist, but search / filter query returns no results, we should probably have something like "No results found. Please try refining your search."

If that will be an issue to separate those out, we can compromise!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7426183).

**Pull request merged**
